### PR TITLE
Completion escape

### LINKED ISSUE #4199: Apostrophes sometimes show as HTML escapes in the completion
When doing `python3 -m qutebrowser --temp-basedir -s aliases '{"x": "session-load -c custom-sess"}'` and doing `:x`, the apostrophes in the description show up as escapes (but that isn't always the case):

image

Maybe Qt interprets it as plaintext instead of rich text?

cc @rcorre if you feel like taking a look

### PR DESCRIPTION
**Pull request by rcorre**
Resolves #4199 (and adds a little test coverage :grin:)

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/4220)

### CODE REVIEW

File: tests/unit/completion/test_completiondelegate.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tests for the CompletionView Object."""
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: Should probably be `CompletionDelegate`, but feel free to remove it completely (I *think* I turned off pylint complaining about that for tests).

File: tests/unit/completion/test_completiondelegate.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+from unittest import mock
+
+import pytest
+
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: Remove this blank line (both pytest and PyQt5 are third-party libs)

File: tests/unit/completion/test_completiondelegate.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+from PyQt5.QtGui import QPainter
+
+from qutebrowser.completion import completiondelegate
+
+
+@pytest.fixture
+def painter():
+    """Create the CompletionView used for testing."""
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: That seems to create a `QPainter`? Seems fine to remove the docstring though.

File: tests/unit/completion/test_completiondelegate.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    return mock.Mock(spec=QPainter)
+
+
+def _qt_mock(klass, mocker):
+    m = mocker.patch(
+        'qutebrowser.completion.completiondelegate.{}'.format(klass),
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: I recently learned about `mocker.patch.object(target, attribute)` which would make this nicer, as you can do `mocker.patch.object(completiondelegate, klass)`.

File: tests/unit/completion/test_completiondelegate.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        autospec=True)
+    return m
+
+
+@pytest.fixture
+def mock_style_option(mocker):
+    """Create the CompletionView used for testing."""
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: Remove :wink:

File: tests/unit/completion/test_completiondelegate.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    return _qt_mock('QStyleOptionViewItem', mocker)
+
+
+@pytest.fixture
+def mock_text_document(mocker):
+    """Create the CompletionView used for testing."""
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: Remove :wink:

File: tests/unit/completion/test_completiondelegate.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+def painter():
+    """Create the CompletionView used for testing."""
+    return mock.Mock(spec=QPainter)
+
+
+def _qt_mock(klass, mocker):
+    m = mocker.patch(
+        'qutebrowser.completion.completiondelegate.{}'.format(klass),
+        autospec=True)
+    return m
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: nitpick: Could probably just return it directly instead of assigning it to a name.

File: tests/unit/completion/test_completiondelegate.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+@pytest.fixture
+def delegate(mock_style_option, mock_text_document, config_stub, mocker, view):
+    _qt_mock('QStyle', mocker)
+    _qt_mock('QTextOption', mocker)
+    _qt_mock('QAbstractTextDocumentLayout', mocker)
+    completiondelegate._cached_stylesheet = mock.Mock()
+    delegate = completiondelegate.CompletionItemDelegate()
+    parent = mock.Mock()
+    parent.return_value = view
+    delegate.parent = parent
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: Not too happy about this `parent` dance - I think it'd be nicer to make a `FakeView` class which inherits `QObject` (and does `self.pattern = None` in `__init__`), make the `view` fixture return that, and then you can just pass that as parent to `CompletionItemDelegate()`.

File: tests/unit/completion/test_completiondelegate.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+def view():
+    return mock.Mock()
+
+
+@pytest.fixture
+def delegate(mock_style_option, mock_text_document, config_stub, mocker, view):
+    _qt_mock('QStyle', mocker)
+    _qt_mock('QTextOption', mocker)
+    _qt_mock('QAbstractTextDocumentLayout', mocker)
+    completiondelegate._cached_stylesheet = mock.Mock()
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: Why is this needed?

File: tests/unit/completion/test_completiondelegate.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    delegate = completiondelegate.CompletionItemDelegate()
+    parent = mock.Mock()
+    parent.return_value = view
+    delegate.parent = parent
+    delegate.initStyleOption = mock.Mock()
+    delegate.setTextDirection = mock.Mock()
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: You can probably pass `spec={}` to lock those down a bit - but can't we just let our code call those?

File: tests/unit/completion/test_completiondelegate.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    ('foo', "'foo'", "'{foo}'"),
+    ('x', "'x'", "'{x}'"),
+    ('lt', "<lt", "&lt;{lt}"),
+])
+def test_paint(delegate, painter, view, mock_style_option, mock_text_document,
+               pat, txt_in, txt_out):
+    """Ensure set_model actually sets the model and expands all categories."""
+    view.pattern = pat
+    mock_style_option().text = txt_in
+    index = mock.Mock()
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: `spec=QModelIndex`?

File: tests/unit/completion/test_completiondelegate.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    parent = mock.Mock()
+    parent.return_value = view
+    delegate.parent = parent
+    delegate.initStyleOption = mock.Mock()
+    delegate.setTextDirection = mock.Mock()
+    return delegate
+
+
+@pytest.mark.parametrize('pat,txt_in,txt_out', [
+    # { and } represent the open/close html tags for highlighting
```
Comment (The-Compiler) [inline tests/unit/completion/test_completiondelegate.py]: What about using `[` and `]` instead? `{foo}` looks a lot like a Python format string, which is a bit confusing.

File: qutebrowser/completion/completiondelegate.py [hunk 13]
```diff
@@ -202,10 +202,16 @@ def _get_textdoc(self, index):
...
+                pat = '({})'.format(re.escape(pattern).replace(r'\ ', r'|'))
+                parts = re.split(pat, self._opt.text, flags=re.IGNORECASE)
+                fmt = '<span class="highlight">{}</span>'
+                escape = lambda s: html.escape(s, quote=False)
+                highlight = lambda s: fmt.format(escape(s))
+                # matches are at every odd index
+                text = ''.join([
+                    highlight(s) if i % 2 == 1 else escape(s)
+                    for i, s in enumerate(parts)
+                ])
```
[the remark this answers is not in the document]
Comment (rcorre) [reply qutebrowser/completion/completiondelegate.py]: > every non-overlapping occurrence of pattern

What about the non-matches, that still have to be `html` escaped?

File: tests/unit/completion/test_completiondelegate.py (hunk 9 above)
Comment (rcorre) [reply tests/unit/completion/test_completiondelegate.py]: `_get_textdoc`, which contains the code I'm trying to test, does:

```
assert _cached_stylesheet is not None
self._doc.setDefaultStyleSheet(_cached_stylesheet)
```

File: tests/unit/completion/test_completiondelegate.py (hunk 10 above)
Comment (rcorre) [reply tests/unit/completion/test_completiondelegate.py]: Without mocking:

```
E       TypeError: initStyleOption(self, QStyleOptionViewItem, QModelIndex): argument 1 has unexpected type 'NonCallableMagicMock'
```

Turns out `setTextDirection` is fine though.

File: tests/unit/completion/test_completiondelegate.py (hunk 9 above)
Comment (rcorre) [reply tests/unit/completion/test_completiondelegate.py]: didn't realize it was just a string argument though, `''` works fine.

File: tests/unit/completion/test_completiondelegate.py (hunk 12 above)
Comment (rcorre) [reply tests/unit/completion/test_completiondelegate.py]: will do. That potential confusion did occur to me. I think I just liked how vim actually highlighted the portions that were supposed to be highligted that way :grin:

**Comment by The-Compiler** [comment]
So I still felt like there should be some easier way to do this, and discovered [QSyntaxHighlighter](http://doc.qt.io/qt-5/qsyntaxhighlighter.html):

```diff
diff --git a/qutebrowser/completion/completiondelegate.py b/qutebrowser/completion/completiondelegate.py
index bc856d1a6..9e64738ff 100644
--- a/qutebrowser/completion/completiondelegate.py
+++ b/qutebrowser/completion/completiondelegate.py
@@ -26,9 +26,10 @@ import re
 import html
 
 from PyQt5.QtWidgets import QStyle, QStyleOptionViewItem, QStyledItemDelegate
-from PyQt5.QtCore import QRectF, QSize, Qt
+from PyQt5.QtCore import QRectF, QSize, Qt, pyqtSlot
 from PyQt5.QtGui import (QIcon, QPalette, QTextDocument, QTextOption,
-                         QAbstractTextDocumentLayout)
+                         QAbstractTextDocumentLayout, QSyntaxHighlighter,
+                         QTextCharFormat)
 
 from qutebrowser.config import config
 from qutebrowser.utils import qtutils, jinja
@@ -37,6 +38,21 @@ from qutebrowser.utils import qtutils, jinja
 _cached_stylesheet = None
 
 
+class _Highlighter(QSyntaxHighlighter):
+
+    def __init__(self, doc, pattern):
+        super().__init__(doc)
+        self._format = QTextCharFormat()
+        self._format.setForeground(Qt.red)  # FIXME
+        self._pattern = pattern
+
+    def highlightBlock(self, text):
+        for match in re.finditer(self._pattern, text):
+            start, end = match.span()
+            length = end - start
+            self.setFormat(start, length, self._format)
+
+
 class CompletionItemDelegate(QStyledItemDelegate):
 
     """Delegate used by CompletionView to draw individual items.
@@ -201,20 +217,10 @@ class CompletionItemDelegate(QStyledItemDelegate):
             view = self.parent()
             pattern = view.pattern
             columns_to_filter = index.model().columns_to_filter(index)
+            self._doc.setPlainText(self._opt.text)
             if index.column() in columns_to_filter and pattern:
-                pat = '({})'.format(re.escape(pattern).replace(r'\ ', r'|'))
-                parts = re.split(pat, self._opt.text, flags=re.IGNORECASE)
-                fmt = '<span class="highlight">{}</span>'
-                escape = lambda s: html.escape(s, quote=False)
-                highlight = lambda s: fmt.format(escape(s))
-                # matches are at every odd index
-                text = ''.join([
-                    highlight(s) if i % 2 == 1 else escape(s)
-                    for i, s in enumerate(parts)
-                ])
-                self._doc.setHtml(text)
-            else:
-                self._doc.setPlainText(self._opt.text)
+                pat = re.escape(pattern).replace(r'\ ', r'|')
+                _Highlighter(self._doc, pat)
         else:
             self._doc.setHtml(
                 '<span style="font: {};">{}</span>'.format(
```

(can probably also get rid of all the `_cached_stylesheet` stuff now that it's not needed anymore)

Not sure how to feel about it. It seems a lot less error-prone (and doesn't involve HTML "serializing" and parsing), but it probably renders all the tests you wrote unusable :worried:  I also have no idea how it looks performance-wise compared to the current code. I hope it'd be a slight improvement, but no clue.

What do you think?

**Comment by rcorre** [comment]
I do think it looks cleaner! I'll see if I can come up with some decent tests around it.

**Comment by The-Compiler** [comment]
FWIW I'm fine with tests which are a bit less isolated and e.g. involve Qt objects, or even Qt widgets.

It's always a trade-off between having slow/unstable tests; but tests which are close to reality and are still usable when there are refactorings. The current end2end tests are definitely too big (and I've realized that in the meantime), but I feel like "pure" unit tests like the ones you added here are perhaps even a bit too much.

In other words, I'd be fine with having tests which involve a bit more Qt foo, maybe even some kind of Qt tree view (or a real `CompletionView`) to test painting the delegate. Would probably also be useful for benchmark tests, too.

File: qutebrowser/completion/completiondelegate.py [hunk 14]
```diff
@@ -194,21 +208,15 @@ def _get_textdoc(self, index):
...
-                repl = r'<span class="highlight">\g<0></span>'
-                pat = html.escape(re.escape(pattern)).replace(r'\ ', r'|')
-                txt = html.escape(self._opt.text)
-                text = re.sub(pat, repl, txt, flags=re.IGNORECASE)
-                self._doc.setHtml(text)
-            else:
-                self._doc.setPlainText(self._opt.text)
+                pat = re.escape(pattern).replace(r'\ ', r'|')
+                _Highlighter(self._doc, pat,
+                             config.val.colors.completion.match.fg)
```
Comment (The-Compiler) [inline qutebrowser/completion/completiondelegate.py]: The type of the setting is currently a `QssColor` which is a string to be used in a stylesheet, but it should now be a `QtColor` instead (which gives you a `QColor` object).

See this failing test: https://travis-ci.org/qutebrowser/qutebrowser/jobs/431903868#L721

Comment (rcorre) [reply qutebrowser/completion/completiondelegate.py]:
```
class QtColor(BaseType):

    """A color value.

    A value can be in one of the following formats:

    * `#RGB`/`#RRGGBB`/`#RRRGGGBBB`/`#RRRRGGGGBBBB`
    * An SVG color name as specified in
      http://www.w3.org/TR/SVG/types.html#ColorKeywords[the W3C specification].
    * transparent (no color)
    """

class QssColor(BaseType):

    """A color value supporting gradients.

    A value can be in one of the following formats:

    * `#RGB`/`#RRGGBB`/`#RRRGGGBBB`/`#RRRRGGGGBBBB`
    * An SVG color name as specified in
      http://www.w3.org/TR/SVG/types.html#ColorKeywords[the W3C specification].
    * transparent (no color)
    * `rgb(r, g, b)` / `rgba(r, g, b, a)` (values 0-255 or percentages)
    * `hsv(h, s, v)` / `hsva(h, s, v, a)` (values 0-255, hue 0-359)
    * A gradient as explained in
      http://doc.qt.io/qt-5/stylesheet-reference.html#list-of-property-types[the Qt documentation] under ``Gradient''
    """
```

Looks like `QssColor` accepts more formats. Do I need to parse it into a `QtColor` to keep backwards compatibility? Or can I just change the setting type and accept that configs using `hsv` or `rgb` for this setting might break?

Comment (The-Compiler) [reply qutebrowser/completion/completiondelegate.py]: Hmm, good point... Gradient's won't work anymore either way, but that's fine.

I'd kinda like to have support for those functions in `QtColor`, also for the additional consistency between `QtColor` and `QssColor`. I suppose it shouldn't be too hard to parse using some regex.

FWIW here's Qt's parsing for it, but it uses their CSS(-ish) parser which isn't exposed: https://github.com/qt/qtbase/blob/v5.11.2/src/gui/text/qcssparser.cpp#L711-L778

Do you want to take a look at that, @rcorre? If you don't have the time I can do so, but I have lots of PRs to look at still :wink:

Comment (rcorre) [reply qutebrowser/completion/completiondelegate.py]: Gotcha, I'll take a look.

**Comment by The-Compiler** [comment]
Thank you! I did a quick profile, and it doesn't seem like a noticable change either way.
